aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -1835,61 +1885,100 @@ Tensor sparse_compressed_to_sparse(const Tensor& self, c10::optional<c10::Layout
       return sparse_compressed_to_sparse_bsr(self, blocksize_, dense_dim_opt);
     }
   case kSparseBsc:
     if (blocksize.has_value()) {
       return sparse_compressed_to_sparse_bsc(self, *blocksize, dense_dim_opt);
     } else {
       const auto blocksize_ = at::sparse_csr::getBlockSize(self);
-      TORCH_CHECK(blocksize_.size() == 2, "to_sparse: ", self.layout(), " to ", layout_,
-                  " conversion requires blocksize specified.");
       return sparse_compressed_to_sparse_bsc(self, blocksize_, dense_dim_opt);
     }
   default:
     break;
   }
-  AT_ERROR("to_sparse: ", self.layout(), " to ", layout_, " conversion not implemented.");
-  return Tensor();
+
+  AT_ERROR("sparse_compressed_to_sparse: ", self.layout(), " to ", layout_to, " conversion not supported");
+  return Tensor{};
 }
 
 Tensor sparse_coo_to_sparse(const Tensor& self, c10::optional<c10::Layout> layout, OptionalIntArrayRef blocksize, c10::optional<int64_t> dense_dim_opt) {
-  Layout layout_ = layout.value_or(kSparse);
-  TORCH_CHECK(!blocksize.has_value() || layout_ == kSparseBsr || layout_ == kSparseBsc,
-              "to_sparse: ", self.layout(), " to ", layout_,
-              " conversion does not use the specified blocksize ", blocksize.value(), ".");
-  if (dense_dim_opt.has_value()) {
-    AT_ERROR("sparse_coo_to_sparse for ", self.layout(), " to ", layout_, " conversion does not support specifying number of dense dimensions");
-  }
-  if (self.layout() == layout_) {
-    return self;
-  }
-  switch (layout_) {
+  auto layout_to = layout.value_or(kSparse);
+  TORCH_CHECK(self.layout() != layout_to, "sparse_coo_to_sparse: same input and output layouts are not supported");
+  _to_sparse_check_arguments("sparse_coo_to_sparse", self, layout_to, blocksize, dense_dim_opt);
+
+  switch (layout_to) {
   case kStrided:
     return self.to_dense();
-  case kSparse:
-    return self;
   case kSparseCsr:
     return self.to_sparse_csr(dense_dim_opt);
   case kSparseCsc:
     return self.to_sparse_csc(dense_dim_opt);
   case kSparseBsr:
-    TORCH_CHECK(blocksize.has_value(), "to_sparse: ", self.layout(), " to ", layout_,
-                " conversion requires blocksize specified.");
     return self.to_sparse_bsr(*blocksize, dense_dim_opt);
   case kSparseBsc:
-    TORCH_CHECK(blocksize.has_value(), "to_sparse: ", self.layout(), " to ", layout_,
-                " conversion requires blocksize specified.");
     return self.to_sparse_bsc(*blocksize, dense_dim_opt);
     default:
       break;
   }
-  AT_ERROR("to_sparse not implemented for ", self.layout(), " to ", *layout, " conversion");
-  return Tensor();
+
+  AT_ERROR("sparse_coo_to_sparse: ", self.layout(), " to ", layout_to, " conversion not supported");
+  return Tensor{};
 }
 
+Tensor to_sparse(const Tensor& self, const int64_t sparse_dim) {
+  auto layout_to = kSparse;
+  if (self.layout() == layout_to) {
+    _to_sparse_check_arguments("to_sparse", self, sparse_dim);
+    return self;
```

Comment:
If `self.sparse_dim() != sparse_dim`, we should not silently return `self` because it does not match with user expectation: `coo.to_sparse(sparse_dim).sparse_dim() == sparse_dim`.

Reply:
An error will be reported in case `self.sparse_dim() != sparse_dim`, it is handled by `_to_sparse_check_arguments()`.  Same for other two comments below.

There is a trade-off here: between readability of the code and having some tests, like the one mentioned above, duplicated.  This PR is gathering all the tests in single place, into `_to_sparse_check_arguments()`.  Namely, I was actually going after enumerating all the combinations of arguments and input/output layouts that are not supported for sparse-to-sparse conversions, primarily for the sake of eventually having these mentioned in the docs.  Eventually, it seemed to me it would be easier to grasp such cases, and maintain all the checks, if refactored into `_to_sparse_check_arguments()` functions.

Reply:
For me, "unsupported" can mean two things: (i) it is not implemented but may be implemented in the future, or (ii) it will never be implemented by design.

In docs, only case (ii) should be reported while case (i) not. This is because the docs report what can be done (or what cannot be done) at the moment of reading the docs and in the future. If we would also report (i) in the docs as unsupported, it would have no value to the users. By default, anything that is not documented is also unsupported.

To resolve this discussion, adding a unittest for the invariant
```
coo.to_sparse(sparse_dim).sparse_dim() == sparse_dim
```
would be sufficient. The test must check that an exception is raised when user provides `sparse_dim` value that would violate the invariant above.

Reply:
Add unittest, and updated PR with all the latest changes.